File: day2.rb
def get_sum_valid_games
	sumIDs = 0
	limits = {"red" => 12, "green" => 13, "blue" => 14}
	File.open("day2Input.txt") do |f|
		f.each_line do |line|
			line.strip!
			validGame = true
			gameID = Integer(line.split(": ")[0].slice(4..))
			game = line.split(": ")[1]
			sets = game.split("; ").each do |set|
				cubes = set.split(", ").each do |cubes|
					parts = cubes.split(" ")
					validCubes = Integer(parts[0]) <= limits[parts[1]]
					validGame = validGame && validCubes
				end
			end
			if validGame
				sumIDs += gameID
			end
		end
	end
	return sumIDs
end

def get_sum_games_power
	sumPowers = 0
	File.open("day2Input.txt") do |f|
		f.each_line do |line|
			line.strip!
			minimums = {"red" => 0, "green" => 0, "blue" => 0}
			game = line.split(": ")[1]
			sets = game.split("; ").each do |set|
				cubes = set.split(", ").each do |cubes|
					parts = cubes.split(" ")
					currentMin = minimums[parts[1]]
					numCubes = Integer(parts[0])
					if numCubes > currentMin
						minimums[parts[1]] = numCubes
					end
				end
			end
			sumPowers += minimums.values.reduce(:*)
		end
	end
	return sumPowers
end

puts get_sum_valid_games
puts get_sum_games_power

File: day02/day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func print_error(err error) {
	if err != nil {
		fmt.Println(err)
	}
}

func get_valid_games() {
	file, err := os.Open("day2Input.txt")
	print_error(err)
	defer file.Close()

	totalValidGames := 0
	scanner := bufio.NewScanner(file)
	cubesMap := map[string]int{"red": 12, "green": 13, "blue": 14}
	for scanner.Scan() {
		line := scanner.Text()
		gameID, err := strconv.Atoi(strings.Split(strings.Split(line, ": ")[0], " ")[1])
		print_error(err)
		sets := strings.Split(strings.Split(line, ": ")[1], "; ")
		validGame := true
		for _, set := range sets {
			cubes := strings.Split(set, ", ")
			for _, cube := range cubes {
				parts := strings.Split(cube, " ")
				numCubes, err := strconv.Atoi(parts[0])
				print_error(err)
				if numCubes > cubesMap[parts[1]] {
					validGame = false
					break
				}
			}
			if !validGame {
				break
			}
		}
		if validGame {
			totalValidGames += gameID
		}
	}
	fmt.Println(totalValidGames)
}

func get_minimum_cube_powers() {
	file, err := os.Open("day2Input.txt")
	print_error(err)
	defer file.Close()

	totalPowers := 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		sets := strings.Split(strings.Split(line, ": ")[1], "; ")
		minimumCubes := map[string]int{"red": 0, "green": 0, "blue": 0}
		for _, set := range sets {
			cubes := strings.Split(set, ", ")
			for _, cube := range cubes {
				parts := strings.Split(cube, " ")
				numCubes, err := strconv.Atoi(parts[0])
				print_error(err)
				if numCubes > minimumCubes[parts[1]] {
					minimumCubes[parts[1]] = numCubes
				}
			}
		}
		power := 1
		for _, v := range minimumCubes {
			power *= v
		}
		totalPowers += power
	}
	fmt.Println(totalPowers)
}

func main() {
	get_valid_games()
	get_minimum_cube_powers()
}